orttraining/orttraining/training_ops/cuda/communication/recv.cc

```diff
@@ -103,13 +104,11 @@ void Recv::ReceiveData(
     // Find the next aligned offset in the tensor buffer to meet alignment requirement
     tensor_offset_in_bytes = GetAggregatedAlignedAddress(tensor_offset_in_bytes);
 
-    // Keep the sync copy in the previous design
-    // TODO they can be moved to async call after global stream becoming accessible
+    // Copy data out from buffer.
     CUDA_CALL(cudaMemcpyAsync(tensor->MutableDataRaw(), buffer.get() + tensor_offset_in_bytes,
```

Comment:
>cudaMemcpyAsync [](start = 14, length = 15)

Is it possible to use output buffer directly in send/recv to avoid this copy?

Reply:
It'd look tricky. The size of buffer is different than the sum of all output sizes. The buffer size is usually larger for putting all communicated tensors in a consecutive memory block.

---
In reply to: [483803797](https://github.com/microsoft/onnxruntime/pull/5054#discussion_r483803797) [](ancestors = 483803797)